api/api.cc

```diff
@@ -233,5 +233,32 @@ future<> set_server_done(http_context& ctx) {
     });
 }
 
+void req_params::process(const request& req) {
+    // Process mandatory parameters
+    for (auto& [name, ent] : params) {
+        if (!ent.is_mandatory) {
+            continue;
+        }
+        try {
+            ent.value = req.param[name];
+        } catch (std::out_of_range&) {
+            throw httpd::bad_param_exception(fmt::format("Mandatory parameter '{}' was not provided", name));
+        }
+    }
+
+    // Process optional parameters
+    for (auto& [name, value] : req.query_parameters) {
+        try {
+            auto& ent = params.at(name);
+            if (ent.is_mandatory) {
+                on_internal_error(apilog, fmt::format("Error while populating request: parameter {} was expected to be mandatory", name));
```

Comment:
I don't understand this case. Why is this an internal error and not a user error?

Reply:
Mandatory parameters are processed separately by the seastar httpd layer and get stored in req.params (in contrast to req.query_paramters that stores the option params).  The are derived from the path url, basically like position command line args.  E.g. `/storage_service/keyspace_scrub/my_keyspace` generates a mandatory param named `keyspace` with value equal to `my_keyspace`.  The name `keyspace` is generated by https://github.com/scylladb/scylla/blob/69fcc053bbb9156b8b88116a8dc8cc976148d86d/api/api-doc/storage_service.json#L801
if I understand correctly.

The idea here was that getting an optional param that is defined as a mandatory param means that it's ill defined, but re-thinking about it, it's possible for the user to pass an optional param giving the same name as the mandatory param, in which case the error should be `httpd::bad_param_exception`.  I agree that it'd be better to presume the latter scenario over the former,

Reply:
Fixed in v2.